src/driver/driver.cpp

```diff
@@ -84,8 +73,14 @@ DriverStatus EvolutionDriver::Execute() {
     pmesh->mbcnt += pmesh->nbtotal;
     pmesh->step_since_lb++;
 
+    // adding several global fences in order to ensure that the Mesh function,
+    // which may operate (or use results from multiple MeshBlocks), access
+    // the MeshBlock data only after all MeshBlock kernels have finished.
+    Kokkos::fence();
     pmesh->LoadBalancingAndAdaptiveMeshRefinement(pinput);
+    Kokkos::fence();
     if (pmesh->modified) InitializeBlockTimeSteps();
+    Kokkos::fence();
     SetGlobalTimeStep();
```

Comment:
Are these fences strictly necessary?
Since these procedures run over all blocks, they shouldn't return in a partially complete state.

Reply:
They may not absolutely be necessary (right now) but also don't hurt performance and will be a safe option for future changes.
Especially, once we think about using `parallel_reduce` with the result stored in a view instead of a scalar (e.g., for calculating the timestep) there's no implicit fence at the end of the `parallel_reduce` anymore. Thus, we need a global fence to ensure that all kernels are done before we do a reduction over the results stored in all MeshBlocks.